Userland/Libraries/LibCrypto/BigInt/Algorithms/ModularPower.cpp

```diff
@@ -49,4 +49,239 @@ void UnsignedBigIntegerAlgorithms::destructive_modular_power_without_allocation(
+    UnsignedBigInteger& temp_for_swap,
+    UnsignedBigInteger& output)
+{
+    VERIFY(modulo.is_odd());
+
+    // Note: While this is a constexpr variable for clarity and could be changed in theory,
+    // various optimized parts of the algorithm rely on this value being exactly 4.
+    constexpr size_t window_size = 4;
+
+    size_t num_words = modulo.trimmed_length();
+    UnsignedBigInteger::Word k = inverse_wrapped(modulo.m_words[0]);
+
+    one.set_to(1);
+
+    // rr = ( 2 ^ (2 * modulo.length() * BITS_IN_WORD) ) % modulo
+    shift_left_by_n_words(one, 2 * num_words, x);
+    divide_without_allocation(x, modulo, temp_z, one, z, zz, temp_for_swap, rr);
+    rr.resize_with_leading_zeros(num_words);
+
+    // x = base [% modulo, if x doesn't already fit in modulo's words]
+    x.set_to(base);
+    if (x.trimmed_length() > num_words)
+        divide_without_allocation(base, modulo, temp_z, one, z, zz, temp_for_swap, x);
+    x.resize_with_leading_zeros(num_words);
+
+    one.set_to(1);
+    one.resize_with_leading_zeros(num_words);
+
+    // Compute the montgomery powers from 0 to 2^window_size. powers[i] = x^i
+    UnsignedBigInteger powers[1 << window_size];
+    almost_montgomery_multiplication_without_allocation(one, rr, modulo, temp_z, k, num_words, powers[0]);
+    almost_montgomery_multiplication_without_allocation(x, rr, modulo, temp_z, k, num_words, powers[1]);
+    for (size_t i = 2; i < (1 << window_size); ++i)
+        almost_montgomery_multiplication_without_allocation(powers[i - 1], powers[1], modulo, temp_z, k, num_words, powers[i]);
+
+    z.set_to(powers[0]);
+    z.resize_with_leading_zeros(num_words);
+    zz.set_to(0);
+    zz.resize_with_leading_zeros(num_words);
+
+    ssize_t exponent_length = exponent.trimmed_length();
+    for (ssize_t word_in_exponent = exponent_length - 1; word_in_exponent >= 0; --word_in_exponent) {
+        UnsignedBigInteger::Word exponent_word = exponent.m_words[word_in_exponent];
+        size_t bit_in_word = 0;
+        while (bit_in_word < UnsignedBigInteger::BITS_IN_WORD) {
+            if (word_in_exponent != exponent_length - 1 || bit_in_word != 0) {
+                almost_montgomery_multiplication_without_allocation(z, z, modulo, temp_z, k, num_words, zz);
+                almost_montgomery_multiplication_without_allocation(zz, zz, modulo, temp_z, k, num_words, z);
+                almost_montgomery_multiplication_without_allocation(z, z, modulo, temp_z, k, num_words, zz);
+                almost_montgomery_multiplication_without_allocation(zz, zz, modulo, temp_z, k, num_words, z);
+            }
+            auto power_index = exponent_word >> (UnsignedBigInteger::BITS_IN_WORD - window_size);
+            auto& power = powers[power_index];
+            almost_montgomery_multiplication_without_allocation(z, power, modulo, temp_z, k, num_words, zz);
+
+            // Swap z and zz
+            temp_for_swap.set_to(z);
+            z.set_to(zz);
+            zz.set_to(temp_for_swap);
```

Comment:
Does our bigint have a move ctor?
Could go about it by doing `swap(z, zz)` if it does, might even help performance on very big bigints by not allocating a temp.